### PR TITLE
Enable the pde-build tests

### PR DESCRIPTION
**Pull request by laeubi**
Currently pde test are not executed what leads to regressions detected too late. This enables the tests based on a profile only executed for verification builds.

Please note I have reorganized things a bit to have two explicit executions for tests while the default is suppressed. I also explicitly **enabled** P2 tests as for me it seems they are all disabled by default.

Locally the `ScriptGenerationTests.testBug238177:873 expected:<4> but was:<5>` fails, I'll submit this anyways here so maybe we can see if it also fails at Jenkins/Github.

This also currently _disables_ all tests that use a delta-pack I plan to look into this in a separate step if this one is submitted.

### CODE REVIEW

File: build/org.eclipse.pde.build.tests/META-INF/p2.inf [hunk 1]
```diff
@@ -0,0 +1,2 @@
+requires.0.namespace = org.eclipse.equinox.p2.iu
```
Comment (laeubi) [inline build/org.eclipse.pde.build.tests/META-INF/p2.inf:1]: This probably is no longer required...

**Comment by laeubi** [comment]
So we have now 3 P2 test failures, @iloveeclipse can you confirm that these are actually run in the ibuild or is my analysis correct that they are actually disabled there?

If they are _disabled_ I would propose to disable them here as well, or maybe even delete them WDYT @akurtakov ?

**Comment by iloveeclipse** [comment]
Jenkins:

Test Result (3 failures )
org.eclipse.pde.build.internal.tests.p2.P2Tests.testBug265564
org.eclipse.pde.build.internal.tests.p2.PublishingTests.testBug283060
org.eclipse.pde.build.internal.tests.p2.PublishingTests.testPublish_Packaging_1

SDK Linux build:

https://download.eclipse.org/eclipse/downloads/drops4/I20231211-0900/testresults/html/org.eclipse.pde.build.tests_ep431I-unit-cen64-gtk3-java17_linux.gtk.x86_64_17.html#P2TestSuite

I see tests executed on SDK build with no fails.

**Comment by iloveeclipse** [comment]
Two from tests fail because some features couldn't be found. I assume the tycho need to include them into the target platform for tests. Third fail is a NPE, haven't checked why.

File: build/org.eclipse.pde.build.tests/pom.xml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+							</dependency-resolution>
+						</configuration>
+					</plugin>
+					<plugin>
+						<groupId>org.eclipse.tycho</groupId>
+						<artifactId>tycho-surefire-plugin</artifactId>
+						<configuration>
+							<product>org.eclipse.sdk.ide</product>
+							<testRuntime>p2Installed</testRuntime>
+							<skipTests>true</skipTests> <!-- this supress the default execution -->
```
Comment (HannesWell) [inline build/org.eclipse.pde.build.tests/pom.xml:54]: Isn't this already inherited from the `skipTests` property defined above?

File: build/org.eclipse.pde.build.tests/pom.xml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	<properties>
+		<!-- This is sued in the ibuilds where the tycho test execution is
+		DISABLED and instead we rely on the test.xml! -->
+		<skipTests>true</skipTests>
+		<ignoredWarnings>${tests.ignoredWarnings}</ignoredWarnings>
+	</properties>
+
+	<profiles>
+		<profile>
+			<id>local-pde-tests</id>
```
Comment (HannesWell) [inline build/org.eclipse.pde.build.tests/pom.xml:25]: Can't we just re-use `build-individual-bundles` here?
```suggestion
			<id>build-individual-bundles</id>
```

File: build/org.eclipse.pde.build.tests/META-INF/p2.inf (hunk 1 above)
Comment (HannesWell) [reply build/org.eclipse.pde.build.tests/META-INF/p2.inf:1]: Can you then test if it still works without?

**Review by HannesWell — COMMENTED** [review]
Thank @laeubi for bringing this up again.


I have to admit that I currently fail to see the difference to https://github.com/eclipse-pde/eclipse.pde/pull/980, which prevents to cycles?

Another point is that it looked like the activated tests only use what have been build with the I-builds (see https://github.com/eclipse-pde/eclipse.pde/pull/984) and if I understand the config right, this is no surprise since the test-runtime is p2 provisioned?

File: build/org.eclipse.pde.build.tests/pom.xml (hunk 2 above)
Comment (laeubi) [reply build/org.eclipse.pde.build.tests/pom.xml:54]: properties can be overridden, explicit configurations not :-)
The property is only for the "usual" execution with the ibuild

File: build/org.eclipse.pde.build.tests/pom.xml (hunk 3 above)
Comment (laeubi) [reply build/org.eclipse.pde.build.tests/pom.xml:25]: Yes we can!

**Comment by laeubi** [comment]
> Another point is that it looked like the activated tests only use what have been build with the I-builds (see https://github.com/eclipse-pde/eclipse.pde/pull/984) and if I understand the config right, this is no surprise since the test-runtime is p2 provisioned?

This is sadly right (but not the conclusion), this is not because we use `p2 provisioned` this only means that P2 is used to build the test runtime (what for example also executes touch points, creates profiles and so on) while otherwise Tycho do the copy itself.

The problem is that currently I use the SDK feature, but as it is not build in this reactor, it contains strict version ranges and therefore the currently build versions are not selected. So I need to somehow convince the p2 director to update the pde(build) features to the build versions ones.

**Comment by HannesWell** [comment]
> he problem is that currently I use the SDK feature, but as it is not build in this reactor, it contains strict version ranges and therefore the currently build versions are not selected. So I need to somehow convince the p2 director to update the pde(build) features to the build versions ones.

Understand.
Looking at the sdk.product and the sdk feature I think the relevant parts boil down to the platform, jdt and pde feature:
grafik

I wonder if we can just list those three explicitly? Which would always pull in the pde feature just build.
Maybe you also need the feature `org.eclipse.equinox.executable` for the launcher?

**Comment by laeubi** [comment]
@HannesWell fell free to checkout the PR and make some test locally, the good thing is its quite easy  to execute, but p2 installed runtime has some caveats, e.g. it complains that the launcher lib is missing, so one needs to include that as well, also the executable has to be part of the (not yet) supported deltapack provisioning and so on ... but I'm currently working on make this allow to use the new director mojo and even tycho-surefire to simply use a (externally) provisioned install.

**Comment by laeubi** [comment]
> I see tests executed on SDK build with no fails.

Alright I was confused by this obfuscated statement here:

https://github.com/eclipse-pde/eclipse.pde/blob/d7a9363601eaa8148af6d67cc89b7af0e67a3f78/build/org.eclipse.pde.build.tests/src/org/eclipse/pde/build/tests/P2TestSuite.java#L30

even though I don't understand why we need the property at all if in the `test.xml` the suites are explicitly mentioned, so they should never gets executed together anyways and even then I don't see the value to have disable this ... anyways it seems both are wanted in one way or the other.

File: build/org.eclipse.pde.build.tests/META-INF/p2.inf (hunk 1 above)
Comment (laeubi) [reply build/org.eclipse.pde.build.tests/META-INF/p2.inf:1]: I have now created
- https://github.com/eclipse-pde/eclipse.pde/pull/1015

as it should actually be fixed at the root and not inside a test.

Comment (HannesWell) [reply build/org.eclipse.pde.build.tests/META-INF/p2.inf:1]: Yes that's right.
IIRC some other tests in PDE and the platform have a similar requirement defined in their pom.
If their tests are changed to be  p2-provisioned it should now not be necessary anymore.

Comment (laeubi) [reply build/org.eclipse.pde.build.tests/META-INF/p2.inf:1]: Actually it should work regardless of the install mode...

File: build/org.eclipse.pde.build.tests/pom.xml (hunk 3 above)
Comment (laeubi) [reply build/org.eclipse.pde.build.tests/pom.xml:25]: done

**Comment by laeubi** [comment]
> Another point is that it looked like the activated tests only use what have been build with the I-builds (see #984)

This is now fixed, I'm using the [new director-mojo](https://github.com/eclipse-tycho/tycho/blob/main/RELEASE_NOTES.md#new-director-mojo) now to first created a provisioned sdk install, the test bundle is the installed on top of it, according to my local testing this then installs the version from the build (if it is newer).

**Comment by laeubi** [comment]
@jukzi I noticed that CodeQL seem to execute the test, is this really required for code analysis or can we have it `-DskipTests` (compile but not execute tests)

**Comment by jukzi** [comment]
> @jukzi I noticed that CodeQL seem to execute the test, is this really required for code analysis or can we have it `-DskipTests` (compile but not execute tests)

isn't taht already done in https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/blame/5fd8db2bf030a053655563e856801255323b0db8/.github/workflows/codeQLworkflow.yml#L86?

**Comment by laeubi** [comment]
> > @jukzi I noticed that CodeQL seem to execute the test, is this really required for code analysis or can we have it `-DskipTests` (compile but not execute tests)
> 
> isn't taht already done in https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/blame/5fd8db2bf030a053655563e856801255323b0db8/.github/workflows/codeQLworkflow.yml#L86?

Yes sorry for confusion, I think the problem is that I explicitly need to enable it need to think how to archive that...

**Comment by laeubi** [comment]
I was now able to replicate the deltapack creation as well, there is [one test left](https://ci.eclipse.org/pde/job/eclipse.pde/job/PR-998/5/testReport/org.eclipse.pde.build.internal.tests.p2/PublishingTests/testBug283060/) that probably needs to be enhanced so it does not run into NPE but explains what it was missing.

**Comment by laeubi** [comment]
I think the failing test wants that the profile is actually named `SDKProfile` I adjusted the build now.

**Comment by laeubi** [comment]
Only usual maven warnings because of http.service

**Comment by HannesWell** [comment]
It's great to have this again. Thank @laeubi for this work.

**Pull request merged**